### PR TITLE
Fix instance joining when leader is not first

### LINKED ISSUE #160: [4pt] Take care about fail on joining instance when stateful failover is enabled and leader was promoted
See https://github.com/tarantool/cartridge/issues/1204

### PR DESCRIPTION
**Pull request by hackallcode**
Closes #160

Cartridge has the bug: it's impossible to join instances to replicaset,
in which leader is not the first instance in failover priority.

To resolve this problem, the most painless way is to edit failover priority
before instances joining. So, after joining, we can change failover priority back.

### CODE REVIEW

File: library/cartridge_edit_topology.py [hunk 1]
```diff
@@ -318,6 +318,38 @@ def get_replicaset_params(replicaset, cluster_replicaset, cluster_instances, all
     return replicaset_params, None
 
 
+def change_failover_priority_if_leader_not_first(replicaset_params, cluster_replicaset, cluster_instances):
+    if any([
+        not replicaset_params,
+        not cluster_replicaset or not cluster_replicaset.get('leader'),
```
Comment (dokshina) [inline library/cartridge_edit_topology.py]: Seems that it can be split into two items, it's `any`.

Comment (dokshina) [reply library/cartridge_edit_topology.py]: Oh, I understood. Maybe `(cluster_replicaset or {}).get('leader')`?

File: library/cartridge_edit_topology.py [hunk 2]
```diff
@@ -318,6 +318,38 @@ def get_replicaset_params(replicaset, cluster_replicaset, cluster_instances, all
...
+        not cluster_instances,
+    ]):
+        return replicaset_params
```
Comment (dokshina) [inline library/cartridge_edit_topology.py:326]: What about using a variable to remove multiline if?

```python
leader_not_set = any([
  ...
])

if leader_not_set:
   return ...
```

File: library/cartridge_edit_topology.py [hunk 3]
```diff
@@ -318,6 +318,38 @@ def get_replicaset_params(replicaset, cluster_replicaset, cluster_instances, all
...
+
+    if replicaset_params.get('uuid') is not None and replicaset_params.get('join_servers'):
```
Comment (dokshina) [inline library/cartridge_edit_topology.py]: Let's use an early return to reduce `if` level.

File: library/cartridge_edit_topology.py [hunk 4]
```diff
@@ -318,6 +318,38 @@ def get_replicaset_params(replicaset, cluster_replicaset, cluster_instances, all
...
+        leader_uuid = cluster_replicaset.get('leader')
+        replicaset_instances = cluster_replicaset['instances']
+        first_instance_uuid = cluster_instances[replicaset_instances[0]].get('uuid')
+
+        if leader_uuid != first_instance_uuid:
```
Comment (dokshina) [inline library/cartridge_edit_topology.py]: The same, early-return.

File: module_utils/helpers.py [hunk 5]
```diff
@@ -661,6 +661,22 @@ def get_topology_checksum(console_sock, stateboard):
...
+        return require('cartridge.failover').get_active_leaders()
+    ''')
+    if err:
+        return err
+
+    if not leaders:
+        return None
+
+    for replicaset_name, replicaset_params in replicasets.items():
+        replicaset_params['leader'] = leaders.get(replicaset_params['uuid'])
```
Comment (dokshina) [inline module_utils/helpers.py]: `leader` -> `leader_uuid`. Cartridge replicaset has `leader` member, and it's a dict.

File: library/cartridge_edit_topology.py [hunk 6]
```diff
@@ -318,6 +318,38 @@ def get_replicaset_params(replicaset, cluster_replicaset, cluster_instances, all
...
+            temp_failover_priority_names = []
+            for instance_name in replicaset_instances:
+                if cluster_instances[instance_name]['uuid'] != leader_uuid:
+                    temp_failover_priority_names.append(instance_name)
+                else:
+                    temp_failover_priority_names = [instance_name] + temp_failover_priority_names
+
+            replicaset_params['failover_priority'] = [
+                cluster_instances[s]['uuid'] for s in temp_failover_priority_names
+            ]
```
Comment (dokshina) [inline library/cartridge_edit_topology.py]: Maybe

```python
replicaset_params['failover_priority'] = [
    cluster_instances[s]['uuid'] for s in replicaset_instances
]
replicaset_params['failover_priority'].remove(leader_uuid)
replicaset_params['failover_priority'] = leader_uuid + replicaset_params['failover_priority']
```

?

File: library/cartridge_edit_topology.py [hunk 7]
```diff
@@ -337,12 +369,46 @@ def get_replicasets_params(
...
+            cluster_instances,
+        )
+
         if replicaset_params is not None:
             replicasets_params.append(replicaset_params)
 
     return replicasets_params, None
 
 
+def get_replicasets_params_for_join_instances(
```
Comment (dokshina) [inline library/cartridge_edit_topology.py]: `for_join_instances` -> `to_join_instances` or `for_instances_joining`

File: module_utils/helpers.py [hunk 8]
```diff
@@ -661,6 +661,22 @@ def get_topology_checksum(console_sock, stateboard):
     ''')
 
 
+def enrich_replicasets_with_leaders(control_console, replicasets):
+    leaders, err = control_console.eval_res_err('''
+        return require('cartridge.failover').get_active_leaders()
```
Comment (dokshina) [inline module_utils/helpers.py:615]: Does it work for any failover mode? (Just a question, I'm not sure)

Comment (hackallcode) [reply module_utils/helpers.py:615]: Yes

**Review by dokshina — DISMISSED** [review]
Bellissimo!

**Review by dokshina — APPROVED** [review]

**Pull request merged**